### PR TITLE
add refactoring and add code action setting support

### PR DESCRIPTION
**Pull request by heejaechang**
there is no refactoring provider yet. just plumbing code to enable it on vscode/lsp.

added a plumbing code for codeaction settings.

one can add in settings.json

"python.refactoring.xxx" or "python.quickfix.xxx" and use CodeActionSettings.ReadOption("xxx") to read value.

added hidden "python.quickfix.addimports" for add import feature.

### CODE REVIEW

File: src/LanguageServer/Impl/LanguageServer.Configuration.cs [hunk 1]
```diff
@@ -65,11 +66,51 @@ public sealed partial class LanguageServer {
...
+
+            void AppendToMap(JToken setting, int prefixLength, Dictionary<string, object> map) {
+                if (setting == null || !setting.HasValues) {
+                    return;
+                }
+
+                foreach (var child in setting) {
+                    if (child is JValue value) {
+                        // there shouldn't be duplicates and prefix must exist. 
+                        // but it is not important enough to throw exception so catch and ignore and move on
```
Comment (bschnurr) [inline src/LanguageServer/Impl/LanguageServer.Configuration.cs]: no longer catching

Comment (heejaechang) [reply src/LanguageServer/Impl/LanguageServer.Configuration.cs]: removed the comment. thank you!

**Comment by heejaechang** [comment]
ping @jakebailey @MikhailArkhipov @bschnurr @judej @RaymonGulati1 can you do code review for this PR?

File: src/LanguageServer/Impl/LanguageServer.Configuration.cs [hunk 2]
```diff
@@ -67,11 +68,50 @@ public sealed partial class LanguageServer {
...
                 _server.DidChangeConfiguration(new DidChangeConfigurationParams { settings = settings }, cancellationToken);
             }
         }
 
+        private void HandleCodeActionsChanges(JToken pythonSection) {
+            var refactoring = new Dictionary<string, object>();
+            var quickfix = new Dictionary<string, object>();
+
+            var refactoringToken = pythonSection["refactoring"];
+            var quickfixToken = pythonSection["quickfix"];
```
Comment (jakebailey) [inline src/LanguageServer/Impl/LanguageServer.Configuration.cs]: `quickfix` or `quickFix`? Where are these settings coming from?

File: src/LanguageServer/Impl/CodeActions/CodeActionSettings.cs [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        public T GetQuickFixOption<T>(string key, T defaultValue) => GetOption(_quickFix, key, defaultValue);
+
+        private T GetOption<T>(IReadOnlyDictionary<string, object> map, string key, T defaultValue) {
+            try {
+                if (map.TryGetValue(key, out var value)) {
+                    return (T)value;
+                }
+
+                return defaultValue;
+            } catch {
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/CodeActions/CodeActionSettings.cs]: Seems a bit excessive, we probably want to know if `map` or `key` is null? Perhaps `if(value is T)`?

File: src/LanguageServer/Impl/LanguageServer.Configuration.cs (hunk 2 above)
Comment (heejaechang) [reply src/LanguageServer/Impl/LanguageServer.Configuration.cs]: it is coming from settings.json.

ex) "python.quickfix.addimport": false

no UI. just a way to pass setting

Comment (jakebailey) [reply src/LanguageServer/Impl/LanguageServer.Configuration.cs]: Right, I was just wondering if @Anapo14 had looked at the naming ahead of time.

File: src/LanguageServer/Impl/CodeActions/CodeActionSettings.cs (hunk 3 above)
Comment (heejaechang) [reply src/LanguageServer/Impl/CodeActions/CodeActionSettings.cs]: want to remove try/catch?

Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/CodeActions/CodeActionSettings.cs]: Generally yet, this appears to be heavy-handed. We we can handle error normally, we should. At worst, code may eat exceptions except critical ones, there is `IsCriticalException()` extension.

File: src/LanguageServer/Test/MissingImportCodeActionTests.cs [hunk 4]
```diff
@@ -503,7 +521,7 @@ import socket
             // calculate actions
             var diagnosticSpan = spans["diagnostic"].First().ToSourceSpan(analysis.Ast);
             var diagnostics = GetDiagnostics(analysis, diagnosticSpan, MissingImportCodeActionProvider.Instance.FixableDiagnostics);
-            var codeActions = await new CodeActionSource(analysis.ExpressionEvaluator.Services).GetCodeActionsAsync(analysis, diagnostics, CancellationToken.None);
+            var codeActions = await new QuickFixCodeActionSource(analysis.ExpressionEvaluator.Services).GetCodeActionsAsync(analysis, CodeActionSettings.Default, diagnostics, CancellationToken.None);
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Test/MissingImportCodeActionTests.cs]: Should probably pass reasonable cancellation token

Comment (MikhailArkhipov) [reply src/LanguageServer/Test/MissingImportCodeActionTests.cs]: Oh, nvm its test code

File: src/LanguageServer/Impl/Sources/RefactoringCodeActionSource.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            ImmutableArray<IRefactoringCodeActionProvider>.Empty;
+
+        private readonly IServiceContainer _services;
+
+        public RefactoringCodeActionSource(IServiceContainer services) {
+            _services = services;
+        }
+
+        public async Task<CodeAction[]> GetCodeActionsAsync(IDocumentAnalysis analysis, CodeActionSettings settings, Range range, CancellationToken cancellationToken) {
+            cancellationToken.ThrowIfCancellationRequested();
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/Sources/RefactoringCodeActionSource.cs]: Might be better to check cancellation inside the loop?

File: src/LanguageServer/Test/MissingImportCodeActionTests.cs (hunk 4 above)
Comment (heejaechang) [reply src/LanguageServer/Test/MissingImportCodeActionTests.cs]: any example on existing test?

File: src/LanguageServer/Impl/Sources/RefactoringCodeActionSource.cs (hunk 5 above)
Comment (heejaechang) [reply src/LanguageServer/Impl/Sources/RefactoringCodeActionSource.cs]: sure

File: src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs [hunk 6]
```diff
@@ -34,9 +34,10 @@ public interface ICodeActionProvider {
         /// Returns <see cref="CodeAction" /> that can potentially fix given diagnostic
         /// </summary>
         /// <param name="analysis"><see cref="IDocumentAnalysis" /> of the file where <paramref name="diagnostic"/> reported</param>
+        /// <param name="settings">settings related to code actions one can query to get user preferences</param>
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs]: `Settings` capitalization

File: src/LanguageServer/Test/MissingImportCodeActionTests.cs (hunk 4 above)
Comment (heejaechang) [reply src/LanguageServer/Test/MissingImportCodeActionTests.cs]: found timeout attribute supported by mstest. made test to use it.

File: src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs [hunk 7]
```diff
@@ -13,17 +13,17 @@
...
 using Microsoft.Python.Core.Collections;
+using Microsoft.Python.LanguageServer.CodeActions;
 using Microsoft.Python.LanguageServer.Protocol;
 
 namespace Microsoft.Python.LanguageServer {
-    public interface ICodeActionProvider {
+    public interface IQuickFixCodeActionProvider {
         /// <summary>
         /// Returns error code this code action can provide fix for. this error code must be same as ones that are reported to host as diagnostics
         /// ex) error code from linter
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs]: `ex)` ? Did you mean `for example`?

File: src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs [hunk 8]
```diff
@@ -13,17 +13,17 @@
...
 using Microsoft.Python.Analysis.Diagnostics;
 using Microsoft.Python.Core.Collections;
+using Microsoft.Python.LanguageServer.CodeActions;
 using Microsoft.Python.LanguageServer.Protocol;
 
 namespace Microsoft.Python.LanguageServer {
-    public interface ICodeActionProvider {
+    public interface IQuickFixCodeActionProvider {
         /// <summary>
         /// Returns error code this code action can provide fix for. this error code must be same as ones that are reported to host as diagnostics
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs]: `This error...` (capitalization)`

File: src/LanguageServer/Impl/Definitions/IRefactoringCodeActionProvider.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+using Microsoft.Python.LanguageServer.CodeActions;
+using Microsoft.Python.LanguageServer.Protocol;
+
+namespace Microsoft.Python.LanguageServer {
+    public interface IRefactoringCodeActionProvider {
+        /// <summary>
+        /// Returns <see cref="CodeAction" /> for the given <paramref name="range"/>. What it would do is up to the refactoring
+        /// </summary>
+        /// <param name="analysis"><see cref="IDocumentAnalysis" /> of the file where <paramref name="range"/> exists</param>
+        /// <param name="settings">settings related to code actions one can query to get user preferences</param>
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/Definitions/IRefactoringCodeActionProvider.cs]: Capitalize first letter in sentences

Comment (MikhailArkhipov) [reply src/LanguageServer/Impl/Definitions/IRefactoringCodeActionProvider.cs]: Also periods at the end of sentences would be nice.

File: src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs (hunk 7 above)
Comment (heejaechang) [reply src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs]: yep.

File: src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs (hunk 8 above)
Comment (heejaechang) [reply src/LanguageServer/Impl/Definitions/IQuickFixCodeActionProvider.cs]: sure.

File: src/LanguageServer/Impl/Definitions/IRefactoringCodeActionProvider.cs (hunk 9 above)
Comment (heejaechang) [reply src/LanguageServer/Impl/Definitions/IRefactoringCodeActionProvider.cs]: done

File: src/LanguageServer/Impl/LanguageServer.Configuration.cs [hunk 10]
```diff
@@ -67,11 +68,50 @@ public sealed partial class LanguageServer {
...
+                    if (child is JValue value) {
+                        // there shouldn't be duplicates and prefix must exist.
+                        var path = child.Path;
+                        if (path.Length <= prefixLength) {
+                            // nothing to add
+                            continue;
+                        }
+
+                        // get rid of common "settings.python..." prefix
+                        map[path.Substring(prefixLength)] = value.Value;
```
Comment (MikhailArkhipov) [inline src/LanguageServer/Impl/LanguageServer.Configuration.cs]: Maybe just pick stuff after the last dot? Up to you.

Comment (heejaechang) [reply src/LanguageServer/Impl/LanguageServer.Configuration.cs]: I would like people to have something like

python.quickfix.addimport.prefer_import = true;
python.quickfix.addimport = true;
python.quickfix.addimport.abbreviation = { "np" : "numpy", "pd": "panda" }

and access them as

settings.GetQuickfixSettings("addimport.prefer_import", false) or
settings.GetQuickfixSettings("addimport", true) or
settings.GetQuickfixSettings("addimport.abbreviation", EmptyDictionary ) 

something like that.

**Comment by MikhailArkhipov** [comment]
LGTM, @jakebailey ?

File: src/LanguageServer/Impl/LanguageServer.Configuration.cs (hunk 2 above)
Comment (jakebailey) [reply src/LanguageServer/Impl/LanguageServer.Configuration.cs]: Per this: https://github.com/microsoft/vscode/blob/434662c6cb11b4cad41bd4bf80c23812aeb319d6/src/vs/editor/contrib/codeAction/types.ts#L13-L19

quickfix is probably fine. I just wanted to ensure we had a standardized name.

File: src/Analysis/Ast/Test/AnalysisTestBase.cs [hunk 11]
```diff
@@ -38,7 +38,9 @@
 
 namespace Microsoft.Python.Analysis.Tests {
     public abstract class AnalysisTestBase {
-        protected TimeSpan AnalysisTimeout { get; set; } = TimeSpan.FromMinutes(1);
+        protected const int AnalysisTimeoutInMS = 1000 * 60;
```
Comment (jakebailey) [inline src/Analysis/Ast/Test/AnalysisTestBase.cs]: Any reason for this change? Seems unrelated (but has the same affect).

**Comment by heejaechang** [comment]
@MikhailArkhipov so you okay except the option name?

File: src/Analysis/Ast/Test/AnalysisTestBase.cs (hunk 11 above)
Comment (heejaechang) [reply src/Analysis/Ast/Test/AnalysisTestBase.cs]: I need the const for attribute. so made the const and made existing one to consume that const

**Comment by heejaechang** [comment]
any feedback from other people?

File: src/Analysis/Ast/Test/AnalysisTestBase.cs (hunk 11 above)
Comment (jakebailey) [reply src/Analysis/Ast/Test/AnalysisTestBase.cs]: Oh, oops. My github diff had the uses of this new constant hidden so I missed it.

**Comment by heejaechang** [comment]
thank you!

**Pull request merged**